src/Metal/MTLArgumentEncoder.cs

```diff
@@ -1,11 +1,54 @@
 using System;
+using System.Runtime.InteropServices;
+using Foundation;
+using ObjCRuntime;
 
 namespace Metal {
+
+#if !XAMCORE_4_0
+	public partial interface IMTLArgumentEncoder {
+		[Obsolete ("Use 'SetManagedBuffers' instead.")]
+		void SetBuffers (IMTLBuffer [] buffers, IntPtr offsets, NSRange range);
+	}
+
+	internal partial class MTLArgumentEncoderWrapper {
+
+		[Obsolete ("Use 'SetManagedBuffers' instead.")]
+		public void SetBuffers (IMTLBuffer [] buffers, IntPtr offsets, NSRange range) 
+			=> throw new NotSupportedException ();
+	}
+#endif
+
 	public static partial class MTLArgumentEncoder_Extensions {
+#if !XAMCORE_4_0
+		[Obsolete ("Use 'SetManagedBuffers' instead.")]
 		public unsafe static void SetBuffers (this IMTLArgumentEncoder This, IMTLBuffer [] buffers, nint [] offsets, Foundation.NSRange range)
+			=> throw new NotSupportedException ();
+#endif
+
+#if XAMCORE_4_0
+		public static void SetBuffers (this IMTLArgumentEncoder table, IMTLBuffer[] buffers, nuint[] offsets, NSRange range)
+#else
+		public static void SetManagedBuffers (this IMTLArgumentEncoder table, IMTLBuffer[] buffers, nuint[] offsets, NSRange range)
+#endif
 		{
-			fixed (void* handle = offsets)
-				This.SetBuffers (buffers, (IntPtr)handle, range);
+			if (buffers == null)
+				throw new ArgumentNullException (nameof (buffers));
+			if (offsets == null)
+				throw new ArgumentNullException (nameof (offsets));
+
+			var bufferPtrArray = buffers.Length <= 1024 ? stackalloc IntPtr[buffers.Length] : new IntPtr [buffers.Length];
+			// get all intptr from the array to pass to the lower level call
+			for (var i = 0; i < buffers.Length; i++) {
+				bufferPtrArray [i] = buffers [i].Handle;
+			}
+
+			unsafe {
+				fixed (void* buffersPtr = bufferPtrArray)
+				fixed (void* offsetsPtr = offsets) { // can use fixed
+					table.SetBuffers ((IntPtr) buffersPtr, (IntPtr) offsetsPtr, range);
+				}
+			}
```

Comment:
I think you need to add this at the end here, otherwise the GC can theoretically collect the array (and everything within):

```csharp
GC.KeepAlive (buffers);
```

Even that isn't bullet proof (another thread can write to the array and change the entries, causing overwritten entries to be collectable), but that seems like something we don't want to support.